power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

```diff
@@ -59,11 +59,24 @@ template <symmetry_tag sym> inline void copy_y_bus(YBus<sym> const& y_bus, Compl
 
 template <symmetry_tag sym>
 inline void calculate_source_result(IdxRange const& sources, Idx bus_number, YBus<sym> const& y_bus,
-                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output) {
+                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
+                                    ComplexValue<sym> const& i_load_gen_bus) {
+    ComplexValue<sym> const i_source_total = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
+    ComplexTensor<sym> y_ref_total{};
+    ComplexValue<sym> i_norton_total{};
     for (Idx const source : sources) {
         ComplexValue<sym> const u_ref{input.source[source]};
         ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
-        output.source[source].i = dot(y_ref, u_ref - output.u[bus_number]);
+        y_ref_total += y_ref;
+        i_norton_total += dot(y_ref, u_ref);
+    }
+
+    for (Idx const source : sources) {
+        ComplexValue<sym> const u_ref{input.source[source]};
+        ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
+        ComplexValue<sym> const i_norton = dot(y_ref, u_ref);
+        ComplexTensor<sym> const y_ref_normalized = y_ref / y_ref_total;
+        output.source[source].i = (i_norton - dot(y_ref_normalized, i_norton_total)) + dot(y_ref_normalized, i_source_total);
         output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
```

Comment:
The whole function:
```suggestion
template <symmetry_tag sym>
inline void calculate_source_result(IdxRange const& sources, Idx bus_number, YBus<sym> const& y_bus,
                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output) {
                                    PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
                                    ComplexValue<sym> const& i_load_gen_bus) {
    ComplexValue<sym> const i_source_total = conj(output.bus_injection[bus_number] / output.u[bus_number]) - i_load_gen_bus;
    std::vector<ComplexTensor<sys>> y_ref_acc(sources.size());
    std::vector<ComplexValue<sym>> i_norton_acc(sources.size());

    std::transform(sources.begin(), sources.end(), y_ref_acc.begin(),
                [&](Idx const source) -> ComplexTensor<sym> {
                    ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
                    return y_ref; 
                });
    std::transform(sources.begin(), sources.end(), i_norton_acc.begin(),
                [&](Idx const source) -> ComplexTensor<sym> {
                    ComplexValue<sym> const u_ref{input.source[source]};
                    ComplexTensor<sym> const y_ref = y_bus.math_model_param().source_param[source];
                    return dot(y_ref, u_ref); 
                });
    ComplexTensor<sym> const y_ref_total_inv = 1.0 / std::accumulate(y_ref_acc.begin(), y_ref_acc.end() - 1, ComplexTensor<sym>{});
    ComplexValue<sym> const i_norton_total = std::accumulate(i_norton_acc.begin(), i_norton_acc.end() - 1, ComplexValue<sym>{});

    for (size_t i = 0; i < sources.size(); ++i) {
        Idx const source = sources[i];
        ComplexTensor<sym> const y_ref_normalized = y_ref_acc[i] * y_ref_total_inv;
        output.source[source].i = (i_norton_acc[i] - dot(y_ref_normalized, i_norton_total)) + dot(y_ref_normalized, i_source_total);
        output.source[source].s = output.u[bus_number] * conj(output.source[source].i);
    }
}
```

Reply:
if possible, please use `std::ranges::transform(vector, target.begin())` instead of `std::transform(vector.begin(), vector.end(), target.begin())` if possible

Reply:
@mgovers @Jerry-Jinfeng-Guo Implemented something along those lines. Thanks for the suggestion.